core/src/main/java/hudson/Util.java

```diff
@@ -595,6 +596,8 @@ public static String ensureEndsWith(@CheckForNull String subject, @CheckForNull
      * @see DigestUtils#md5Hex(InputStream)
      */
     @Nonnull
+    @SuppressFBWarnings(value = "WEAK_MESSAGE_DIGEST_MD5", justification =
+            "This method should only be used for non-security applications where their weakness is not a problem.")
```

Comment:
this is wrong.  
Whilst "this method should only be used for non security calls" supressing it here will mask those that do make security calls.

Reply:
This really highlights the difficulties of adding something like findsecbugs into existing, legacy systems like Jenkins. What we really should do is to fix everything so that there are no usages of weak message digests like MD5. That's a large project that requires very focused work and consideration of migration or incompatibility. Making the introduction of findsecbugs dependent on that change means in practice that findsecbugs will never be introduced and new code will not be checked. Not introducing findsecbugs makes it less likely that any usages will be checked and any improper usages corrected. Tagging it with the `@SuppressFBWarnings` at least points out that it **shouldn't** be used for security.

Not suppressing it here (i.e., not using findsecbugs) only results in better masking it for anything that already misuses this to make security calls or who might in the future.

We could improve the message somehow. I'm not sure what it would say better.

Reply:
> What we really should do is to fix everything so that there are no usages of weak message digests like MD5

There is nothing wrong with MD5 when used correctly.
using it for a file checksum to check a file is downloaded from a HTTP**S** source and has not been corrupted (because the IP layers will not detect more than a small level of corruption) is perfectly acceptable.  Sure we could go on and use something more secure - but my point is that it is the caller here and this code is correct when used in the previous scenario but NOT when used to check some security related things.  Suppressing this means you will never find those.

The only thing that is safe to do is to remove all helper methods and duplicate code.  this seems bad.

Reply:
No, it's not all that bad. It's been common knowledge for a long time that Jenkins uses MD5 and that MD5 has serious flaws. It's suitable for some applications, but not for cryptographic use. It might be okay for fingerprinting but even there we would be better off moving to SHA-2. Our use of MD5 is certainly not good but no one has said it's really bad and needs immediate correction.

It would be really bad if we wanted to support certain high-security environments, such as FIPS. That's not something we're actively pursuing.

Our options now are basically:
1) Close this PR without merging. Nothing changes. We never find the potential problems.
2) Suppress it for now and move forward. Try to fix code and remove suppressions when we can.
3) Replace all of the MD5 usages in Jenkins that need it. (It would be simplest to replace all of them than to analyze which ones might be okay. Besides FIPS requires it.) Introduce findsecbugs only after all of those are fixed. This option is essentially the same as `#1`, because replacing MD5 isn't going to happen while this PR is relevant.
4) Move the suppression into the exclude file.

`#2` is the best option at this point. Suppressing it gives us a (slightly) better chance of getting around to fixing this because it leaves a marker of the issue. Suppressing it leaves a note that other developers _might_ see helping them know not to use this method or to move away from its use.

Reply:
just to be clear I am not blocking this PR.  just pointing out potential issues in that if we go with this with the say Jenkins is written and will be written in the future we will (I beleive) end up with a huge amount of suppressions just to keep the tool happy and I have a doubt over if it will prevent issues in new code.

> Suppress it for now and move forward. Try to fix code and remove suppressions when we can.

you can't easily do this given that it is a public method used by plugins, you can;t just change that contract unless you want to fix all callers in plugins, or we start actually breaking things intentionally in the name of progress (but I am still missing that direction)

Reply:
Yes, someday if we want to fix this specific issue, use of MD5, we will probably need to break things. If we want to someday support secure environments like FIPS, we certainly will. According to [Wikipedia](https://en.wikipedia.org/wiki/MD5), "As of 2019, MD5 continues to be widely used, in spite of its well-documented weaknesses and deprecation by security experts." I guess that means we're certainly not alone is continuing to use it at least. Small consolation I suppose.

I definitely find that tools like spotbugs helps produce cleaner code and help identify and prevent potential problems. I've found value from the findings generated by findsecbugs. I just wish it had been in place much earlier to point out some things that could have been improved much easier earlier.

Thanks for the discussion.